// @universal-robots/urcap-utils/bin/lib/urservice-helper.js
const fetch = require("node-fetch");
const FormData = require("form-data");
const fs = require("fs");

const BASE_URL = "universal-robots/urservice/api/v1/urcaps";

class UrcapInstallationError extends Error {
  constructor(message, details) {
    super(message);
    this.details = details;
  }
}

function getHost() {
  const args = process.argv.slice(2);
  let idx = args.indexOf("--host");
  let host = "localhost";
  if (process.env.DEV_CONTAINER == "true") {
    host = "host.docker.internal";
  }
  if (idx > -1 && args.length >= idx) {
    // User has specified a host
    host = args[idx + 1];
  }
  let port = "80";
  idx = args.indexOf("--port");
  if (idx > -1 && args.length >= idx) {
    port = args[idx + 1];
  }
  return `${host}:${port}`;
}

function getBaseUrl() {
  return `http://${getHost()}/${BASE_URL}`;
}

async function hostReachable() {
    let response;
    try {
        response = await fetch(`${getBaseUrl()}`)
    } catch (reason) {
        throw new UrcapInstallationError('Host unreachable', reason)
    }
    if (response.status === 200) {
        return true
    } else {
        throw new UrcapInstallationError(response.status, response.statusText)
    }
}

async function urcapExists(vendorId, urcapId) {
    if (await hostReachable()) {
        return fetch(`${getBaseUrl()}/${vendorId}/${urcapId}`).then(
            (response) => {
                return Promise.resolve(response.ok);
            },
            () => {
                throw new UrcapInstallationError("Unable to check if URCap exists");
            },
        );
    }
}

function deleteUrcap(vendorId, urcapId) {
  return fetch(`${getBaseUrl()}/${vendorId}/${urcapId}`, {
    method: "DELETE",
  }).then(
    (response) => {
      if (response.ok) {
        return Promise.resolve("URCap succesfully deleted");
      } else {
        return Promise.reject(
          new Error(
            `Unable to delete URCap: HTTP status code ${response.status}`,
          ),
        );
      }
    },
    (reason) => {
      const details = reason.hasOwnProperty("code") ? reason.code : "N/A";
      throw new UrcapInstallationError("Unable to delete URCap", details);
    },
  );
}

function installOrUpdateUrcap(
  urcapxFile,
  vendorId,
  urcapId,
  forceReinstall = false,
) {
  return new Promise((resolve, reject) => {
    urcapExists(vendorId, urcapId)
      .then((exists) => {
        let method = "POST";
        if (exists) {
          if (forceReinstall) {
            deleteUrcap(vendorId, urcapId).then(
              (deleteMessage) => {
                console.log(deleteMessage);
              },
              (reason) => {
                const details = reason.hasOwnProperty("code")
                  ? reason.code
                  : "N/A";
                throw new UrcapInstallationError(
                  "Unable to delete previous version of URCap",
                  details,
                );
              },
            );
          } else {
            method = "PUT";
          }
        }
        const fileName = urcapxFile.substring(urcapxFile.lastIndexOf("/") + 1);
        const formData = new FormData();
        formData.append("urcapxFile", fs.createReadStream(urcapxFile));
        formData.append("fileName", fileName);
        formData.append("type", "*/*");

        const options = {
          method: method,
          body: formData,
          headers: formData.getHeaders(),
        };

        const requestPromise = fetch(getBaseUrl(), options);
        const jsonPromise = requestPromise.then((response) => response.json());
        return Promise.all([requestPromise, jsonPromise]);
      })
      .then((results) => {
        const response = results[0];
        const json = results[1];
        if (response.ok) {
          resolve(json);
        } else {
          let details = getErrorMessageFromInstallErrorResponse(json);
          throw new UrcapInstallationError(json.statusCode,details);
        }
      })
      .catch((error) => {
        return reject(error);
      });
  });
}

function getErrorMessageFromInstallErrorResponse(json) {
    let completeMessage = "";
    if (json.hasOwnProperty('errors')) {
        let errors = json.errors;
        for (let i = 0; i < errors.length; i++) {
            let error = errors[i]
            if (error.hasOwnProperty('message')) {
                if (i !== 0) {
                    completeMessage += ". "
                }
                completeMessage += error.message
            }
        }
    }
    return completeMessage;
}

module.exports = { deleteUrcap, installOrUpdateUrcap };
